### PR TITLE
USM : TLS connection classification

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

Classify the connection if it contain TLS protocol and add it to the protocol stack and connection tags

```yaml
system_probe_config:
    disable_tcp: false

network_config:
    enable_protocol_classification: true
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/tracer-stats.h
Comment (guyarb) [inline pkg/network/ebpf/c/tracer-stats.h:55]:
```suggestion
static __always_inline bool is_tls_helper(conn_tuple_t *t) {
    return bpf_map_lookup_elem(&tls_connection, t) != NULL;
}
```

Comment (guyarb) [reply pkg/network/ebpf/c/tracer-stats.h:55]: 1. the kernel has a naming convention of `lower_case_with_underscore`
2. nit change

File: pkg/network/ebpf/c/tracer-stats.h
Comment (guyarb) [inline pkg/network/ebpf/c/tracer-stats.h:50]: let's add documentation for this function and the other function, so it will be clear what each of them is doing

File: pkg/network/ebpf/c/tracer-stats.h
Comment (guyarb) [inline pkg/network/ebpf/c/tracer-stats.h:57]:
```suggestion
static __always_inline bool is_tls(conn_tuple_t *t) {
```

File: pkg/network/encoding/encoding_test.go
Comment (guyarb) [inline pkg/network/encoding/encoding_test.go:138]: let's convert `1` into the named const

File: pkg/network/encoding/encoding_test.go
Comment (guyarb) [inline pkg/network/encoding/encoding_test.go:207]: let's use named const instead of `1`

File: pkg/network/event_usm_common.go
Comment (guyarb) [inline pkg/network/event_usm_common.go:14]: where is it being used?

File: pkg/network/tags_linux.go
Comment (guyarb) [inline pkg/network/tags_linux.go:34]: you should use the old method as well
the `ConnTagTLS` is being set if we are capturing the handshake, while the other tags can appear if we have uprobes although we didn't capture the tls handshake

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/tls.h:31]:
```suggestion
static __always_inline bool is_valid_tls_app(u8 app) {
```

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/tls.h:35]:
```suggestion
static __always_inline bool is_valid_tls_version(__u16 version) {
```

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/tls.h:50]:
```suggestion
static __always_inline bool is_payload_length_valid(__u8 app, __u16 tls_len, __u32 buf_size) {
    /* check only for application data layer */
    if (app != TLS_APPLICATION_DATA) {
        return false;
    }

    return buf_size < (sizeof(tls_record_t)+tls_len);
}
```

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/tls.h:70]:
```suggestion
static __always_inline bool is_tls(const char* buf, __u32 buf_size) {
    if (buf_size < TLS_HEADER_SIZE) {
        return false;
    }

    tls_record_t *tls_record = (tls_record_t *)buf;
    if (!is_valid_tls_app(tls_record->app)) {
        return false;
    }

    if (!is_valid_tls_version(tls_record->version)) {
        return false;
    }
        
    if (tls_record->length > MAX_TLS_FRAGMENT_LENGTH) {
        return false;
    }
    
    return is_payload_length_valid(tls_record->app, tls_record->length, buf_size);
}
```

File: pkg/network/ebpf/c/protocols/classification/protocol-classification.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/classification/protocol-classification.h:95]: do that in `clean_protocol_classification`

**Review by guyarb — CHANGES_REQUESTED** [review]

**Comment by guyarb** [comment]
@p-lambert please review as well

File: pkg/network/ebpf/c/protocols/classification/dispatcher-helpers.h
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/classification/dispatcher-helpers.h:14]: I forgot to removed it, thanks

File: pkg/network/ebpf/c/tracer-stats.h
Comment (nplanel) [reply pkg/network/ebpf/c/tracer-stats.h:57]: done

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/tls/tls.h:31]: done

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/tls/tls.h:35]: done

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/tls/tls.h:50]: the compiler SSA will take care of that and this will avoid confusion if something else need to be updated/added

File: pkg/network/ebpf/c/protocols/classification/protocol-classification.h
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/classification/protocol-classification.h:95]: done

File: pkg/network/ebpf/c/tracer-stats.h
Comment (nplanel) [reply pkg/network/ebpf/c/tracer-stats.h:55]: done

File: pkg/network/ebpf/c/tracer-stats.h
Comment (nplanel) [reply pkg/network/ebpf/c/tracer-stats.h:50]: done

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/tls/tls.h:70]: the compiler SSA will take care of that and this will avoid confusion if something else need to be updated/added

File: pkg/network/ebpf/c/tracer-stats.h
Comment (nplanel) [reply pkg/network/ebpf/c/tracer-stats.h]: done

**Review by hmahmood — APPROVED** [review]

File: pkg/network/event_usm_common.go
Comment (nplanel) [reply pkg/network/event_usm_common.go:14]: old things, thanks

File: pkg/network/tags_linux.go
Comment (nplanel) [reply pkg/network/tags_linux.go:34]: TagTLS don't detect handshake only but I can add all tag to be in this case to be on the safe side

File: pkg/network/ebpf/c/tracer-stats.h
Comment (nplanel) [reply pkg/network/ebpf/c/tracer-stats.h:131]: thanks

**Review by ISauve — APPROVED** [review]
shallow review for the ebpf-platform file; lgtm!

File: pkg/network/protocols/tls/openssl.go
Comment (guyarb) [inline pkg/network/protocols/tls/openssl.go:30]: `args` is not used
```suggestion
func RunClientOpenssl(t *testing.T, addr string, port string) {
```

File: pkg/network/protocols/tls/openssl.go
Comment (guyarb) [inline pkg/network/protocols/tls/openssl.go:27]: let's increase the timeout, the CI tends to be slow, so short timeout can lead to flakiness

File: tasks/system_probe.py
Comment (guyarb) [inline tasks/system_probe.py:1423]: no need, it appears in the docker-compose (`pkg/network/protocols/tls/testdata/docker-compose.yml`) already

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1128]:
```suggestion
```

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1229]: there's a huge duplication here, let's simplify it
an example in kafka classification tests (https://github.com/DataDog/datadog-agent/blob/main/pkg/network/tracer/tracer_classification_test.go#L372-L415) where we are running the same test for different kafka versions

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/tls/tls.h
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/tls/tls.h:14]: done

File: pkg/network/ebpf/c/protocols/classification/protocol-classification.h
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/classification/protocol-classification.h:93]: done

File: pkg/network/protocols/tls/openssl.go
Comment (nplanel) [reply pkg/network/protocols/tls/openssl.go:27]: done

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:1229]: I not sure there are a value to factorize tests as in some case it make it more complicated to what was failing and why not the previous one. This could be counter intuitive too
I did factorize for the different tls version but still need to add exception for the last test

File: pkg/network/protocols/testutil/serverutils.go
Comment (guyarb) [inline pkg/network/protocols/testutil/serverutils.go:60]: why do we see that change? it is already on `main` branch https://github.com/DataDog/datadog-agent/blob/main/pkg/network/protocols/testutil/serverutils.go#L27

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1134]: nit: 
```suggestion
			preTracerSetup: nil,
```

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1160]: nit: (unless we parameterized the port)
```suggestion
				}, 4*time.Second, time.Second, "couldn't find TLS connection matching: dstport 44330")
```

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [reply pkg/network/tracer/tracer_usm_linux_test.go]: looks much better now

File: pkg/network/protocols/testutil/serverutils.go
Comment (nplanel) [reply pkg/network/protocols/testutil/serverutils.go:60]: I rebased it

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:1153]: it's not possible this way because I check the client and server behave _correctly_ (scope connection)
client waiting for "Verify return code" and server "exited with code"  (the server exit after 1 accept)

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go]: done

File: pkg/network/protocols/testutil/serverutils.go
Comment (nplanel) [reply pkg/network/protocols/testutil/serverutils.go:72]: this is RunHostServer()

File: pkg/network/ebpf/c/tracer-stats.h
Comment (p-lambert) [inline pkg/network/ebpf/c/tracer-stats.h:55]: nit: given that the map value is a boolean, I'd probably change this to:
```
bool *is_tls = bpf_map_lookup_elem(&tls_connection, t);
return (is_tls != NULL && *is_tls);
```

File: pkg/network/protocols/testutil/serverutils.go
Comment (guyarb) [inline pkg/network/protocols/testutil/serverutils.go:93]: Bryce changed that behavior in RunDokcerServer, to avoid goroutine
make the adjustments to your code as well

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:229]:
```suggestion
	fentryTracerEnabled := tr.ebpfTracer.Type() == connection.EBPFFentry
```

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:278]:
```suggestion
			// socket filter is not supported on fentry tracer
			if foundPathAndHTTPTag {
			        if fentryTracerEnabled {
				        // so we return early if the test was successful until now
				        return true
			        }
				for _, c := range payload.Conns {
					if c.SPort == key.SrcPort && c.DPort == key.DstPort && isTLSTag(c.StaticTags) {
						return true
					}
				}
				t.Logf("HTTP connection %v doesn't contain ConnTagTLS\n", key)
			}
```

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1151]:
```suggestion
				go func() {
					defer wg.Done()
					time.Sleep(5 * time.Second)
					clientSuccess = prototls.RunClientOpenssl(t, "localhost", "44330", tlsVersion)
				}()
```

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [reply pkg/network/tracer/tracer_usm_linux_test.go:1156]: the problem is that you are making a flaky test, if it takes to the server more than 5 seconds to run locally

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1126]: not used in the tests

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1181]: if you have USM, then `EnableHTTPMonitoring` is true
so what is the purpose of the code?

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:1234]: adjust your code to fit Bryce's change https://github.com/DataDog/datadog-agent/pull/15970

**Review by guyarb — CHANGES_REQUESTED** [review]
you have failed tests in the CI

**Comment by nplanel** [comment]
> you have failed tests in the CI
If you have idea which test doesn't cleanup properly

       [4.14.262-200.489.amzn2.x86_64] Error:      	Received unexpected error:
       [4.14.262-200.489.amzn2.x86_64] error guessing offsets: could not start offset ebpf manager: probes activation validation failed: 1 error occurred:
       [4.14.262-200.489.amzn2.x86_64] * {UID:offset EBPFFuncName:tracepoint__net__net_dev_queue EBPFSection:}: couldn't set perf event bpf {UID:offset EBPFFuncName:tracepoint__net__net_dev_queue EBPFSection:}: error attaching bpf program to perf event: file exists

File: pkg/network/protocols/testutil/serverutils.go
Comment (nplanel) [reply pkg/network/protocols/testutil/serverutils.go:93]: done

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:229]: done

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:278]: done

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:1151]: done

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go]: done

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:1173]: The test running at the end to check if we leak connection port (but we can remove the test as it doesn't prove much as it's not the same src port)

USM is not enabled, only
```
	cfg.ProtocolClassificationEnabled = true
	cfg.CollectTCPConns = true

```
to enable socket classifier, but I think we should change the logic and would require cfg.ServiceMonitoringEnabled == true to setup the protocol classifier, WDYT ?

**Review by p-lambert — APPROVED** [review]
LGTM once CI is green though a good amount of the code will have be refactored (and hopefully simplified) once I merge https://github.com/DataDog/datadog-agent/pull/16167

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:272]: consider the following simplification:
```suggestion
			if key.Path.Content == "/200/foobar" && (statsTags == network.ConnTagGnuTLS || statsTags == network.ConnTagOpenSSL) {
				t.Logf("found tag 0x%x %s", statsTags, network.GetStaticTags(statsTags))
				// socket filter is not supported on fentry tracer
				if fentryTracerEnabled {
					// so we return early if the test was successful until now
					return true
				}

				for _, c := range payload.Conns {
					if c.SPort == key.SrcPort && c.DPort == key.DstPort && isTLSTag(c.StaticTags) {
						return true
					}
				}
				t.Logf("HTTP connection %v doesn't contain ConnTagTLS\n", key)
		}
```

**Review by guyarb — APPROVED** [review]
overall looks good
left couple of suggestions, non are a blocker

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:801]: I merged the test with the previous one

File: pkg/network/tracer/tracer_usm_linux_test.go
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:272]: done

**Pull request merged**
